### PR TITLE
suse: dynamic distribution discovery

### PR DESCRIPTION
**Pull request by crozzy**
Currently Suse distributions are predefined in the code, this change adds dynamic support for two Suse distro flavors:
suse.linux.enterprise.server and opensuse.leap.

This change is purposefully picky, it only considers versions:
```
suse.linux.enterprise.server.11
suse.linux.enterprise.server.12
suse.linux.enterprise.server.15

opensuse.leap.15.5
opensuse.leap.15.6
```

But should expand over time

### CODE REVIEW

File: suse/factory.go
Comment (RTann) [inline suse/factory.go:83]: nit: `fc := func(n *html.Node) error { ... }`

File: suse/factory.go
Comment (RTann) [inline suse/factory.go:79]: I am pretty thankful for https://security.access.redhat.com/data/oval/v2/PULP_MANIFEST when I look at the alternative below 😂

File: suse/factory.go
Comment (RTann) [inline suse/factory.go:93]: from StackRox perspective, this is fine, but from Clair's perspective, is it worth making this some kind of if/else or doing some kind of prefix match first? I know regexp evaluation is not exactly performant. Not sure how much that matters, though, nor how much of a perf hit it'd be (probably not a lot?)

File: suse/factory.go
Comment (RTann) [inline suse/factory.go:103]:
```suggestion
						return fmt.Errorf("failed to parse uri %v: %w", uri, err)
```

something like that?

File: suse/factory.go
Comment (RTann) [inline suse/factory.go:114]: ah, is this why it's declared above? never really made a recursive closure like this in Go before

Comment (RTann) [inline suse/factory.go:114]: this error may be overwritten and ignored if this child fails but the next succeeds. Is that desired or should we catch the error?

File: go.mod
Comment (crozzy) [reply go.mod:62]: I mean it comes from this import `golang.org/x/net/html` do you mean a mistake that it's indirect?

File: suse/factory.go
Comment (crozzy) [reply suse/factory.go]: Yeah, it needs to be predefined I believe, I rewrote it as a separate function to be more clear

File: suse/factory.go
Comment (crozzy) [reply suse/factory.go]: This is handled differently with the separate function

File: suse/factory.go
Comment (crozzy) [reply suse/factory.go]: I think this is fine given that:
* This is a "background" process
* The regex is defined previously
* The regex is providing 2 functions, matching and returning relevant captured groups

File: go.mod
Comment (crozzy) [reply go.mod]: I think it's how you'd expect now

File: pkg/ovalutil/pool.go
Comment (RTann) [inline pkg/ovalutil/pool.go:12]: just curious: what is the purpose of these pools? If multiple updaters update OVAL data, wouldn't there be a chance  something is accidentally overwritten?

File: suse/factory.go
Comment (RTann) [inline suse/factory.go:25]: does this add a lot more rows? Do you think we should go back to the bzip2 files and just accept SUSE 11 is no longer supported (or just make a workaround for it)?

File: suse/factory.go
Comment (RTann) [inline suse/factory.go:54]: I don't think this is ever set? Also, I didn't see an etag returned when I curl'd an example endpoint

File: suse/distributionscanner.go
Comment (RTann) [inline suse/distributionscanner.go:120]: do we want to bother verifying the CPE is valid, too, like checking for `part == o` or nah?

File: suse/distributionscanner.go
Comment (RTann) [inline suse/distributionscanner.go:129]: should we just use the version from the CPE?

**Review by RTann — COMMENTED** [review]
a few more things, then should be good to go soon

File: pkg/ovalutil/pool.go
Comment (crozzy) [reply pkg/ovalutil/pool.go:12]: From the docs:
```
A Pool is safe for use by multiple goroutines simultaneously.

Pool's purpose is to cache allocated but unused items for later reuse, relieving pressure on the garbage collector. That is, it makes it easy to build efficient, thread-safe free lists. However, it is not suitable for all free lists. 
```
In this case an updater can grab a reader, do what it wants and return it to the pool, as long as the returning happens after the data is read there isn't a way to overwrite the data. Once the reader is returned it's reset and ready for the next client/updater/goroutine to use.

File: suse/factory.go
Comment (crozzy) [reply suse/factory.go:25]: You can see from the PR desc that it does add some earlier opensuse LEAP versions. I added a minimum LEAP version, now the updates are only concerned with:
```
suse.linux.enterprise.server.11
suse.linux.enterprise.server.12
suse.linux.enterprise.server.15
opensuse.leap.15.5
opensuse.leap.15.6
```
The 3 supported ELS versions and the 2 [LEAP supported versions](https://en.opensuse.org/Lifetime).

File: suse/factory.go
Comment (crozzy) [reply suse/factory.go:54]: Fair enough, looks like the initial updaterset creation cannot rely on not-modifed directives, luckily the OVAL files themselves include etags.

File: suse/distributionscanner.go
Comment (crozzy) [reply suse/distributionscanner.go:120]: We are checking if it's a valid CPE, I think that is enough

File: suse/factory.go
Comment (RTann) [reply suse/factory.go:25]: oh I didn't notice you edited it. I thought it was the same description as before 😂

File: suse/distributionscanner.go
Comment (crozzy) [reply suse/distributionscanner.go:129]: Yeah, that'd probably be cleaner

File: suse/factory.go
Comment (crozzy) [reply suse/factory.go]: updated

File: suse/distributionscanner.go
Comment (RTann) [inline suse/distributionscanner.go:32]: are these used anywhere? sorry just noticed this

File: suse/distributionscanner.go
Comment (RTann) [inline suse/distributionscanner.go:18]: also just realized this: just checking: no need to update the version, right? I don't think anything noticeable to end users changed here

File: suse/factory.go
Comment (RTann) [inline suse/factory.go:25]: good for now. Maybe add a note to revisit this when/if 20 comes out

File: suse/factory.go
Comment (RTann) [inline suse/factory.go:52]: also just noticed this, sorry. What is this protecting? I'm wondering if we have to lock around the HTTP call or if this can be done later

File: suse/distributionscanner.go
Comment (RTann) [inline suse/distributionscanner.go:119]: remove before merge?

File: suse/distributionscanner.go
Comment (crozzy) [reply suse/distributionscanner.go:18]: yeah, the indexing should be the same I believe

File: suse/distributionscanner.go
Comment (crozzy) [reply suse/distributionscanner.go:32]: yeah, they are the files we check for during the indexing

File: suse/factory.go
Comment (crozzy) [reply suse/factory.go:52]: I believe this was ensuring that multiple calls to UpdaterSet were protecting setting the fields belonging to the Factory, but we're no longer saving etag stuff there I think it's ok to remove

File: suse/distributionscanner.go
Comment (crozzy) [reply suse/distributionscanner.go]: :facepalm:

File: suse/distributionscanner.go
Comment (RTann) [inline suse/distributionscanner.go:94]: `return nil` here, too? I guess the next line will cover it, too, though?

File: suse/distributionscanner.go
Comment (RTann) [inline suse/distributionscanner.go:117]: nit: can make the CPEs globals so they're only evaluated once. Haven't looked into the function to see if it saves a ton of work or not, though

File: suse/factory.go
Comment (RTann) [inline suse/factory.go:68]: just checking if this is preferred vs a normal error?

**Review by RTann — DISMISSED** [review]
A few tiny things, but I think should be good to go after this. Let me know if my approval is dismissed

File: suse/distributionscanner.go
Comment (crozzy) [reply suse/distributionscanner.go:117]: My instinct is that any minimal efficiency gain would be lost in legibility:

```go
var (
	openSUSECPEVal, leapCPEVal, suseCPEVal, slesCPEVal cpe.Value
)

func init() {
	var err error
	openSUSECPEVal, err = cpe.NewValue("opensuse")
	if err != nil {
		panic(err)
	}
	leapCPEVal, err = cpe.NewValue("leap")
	if err != nil {
		panic(err)
	}
	suseCPEVal, err = cpe.NewValue("suse")
	if err != nil {
		panic(err)
	}
	slesCPEVal, err = cpe.NewValue("sles")
	if err != nil {
		panic(err)
	}
}

func cpeToDist(r cpe.WFN) (*claircore.Distribution, error) {
	if r.Attr[cpe.Vendor] == openSUSECPEVal {
		if r.Attr[cpe.Product] == leapCPEVal {
			return mkLeapDist(r.String(), r.Attr[cpe.Version].String()), nil
		}
	}
...
```
what do you think?

Comment (RTann) [reply suse/distributionscanner.go:118]: fine as-is 👍

**Review by RTann — APPROVED** [review]

**Pull request merged**
